// node-opcua-pki/dist/lib/toolbox/with_openssl/toolbox.js
"use strict";
/* eslint-disable @typescript-eslint/no-explicit-any */
// ---------------------------------------------------------------------------------------------------------------------
// node-opcua-pki
// ---------------------------------------------------------------------------------------------------------------------
// Copyright (c) 2014-2022 - Etienne Rossignon - etienne.rossignon (at) gadz.org
// Copyright (c) 2022-2024 - Sterfive.com
// ---------------------------------------------------------------------------------------------------------------------
//
// This  project is licensed under the terms of the MIT license.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so,  subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
// Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
// ---------------------------------------------------------------------------------------------------------------------
// tslint:disable:no-console
// tslint:disable:no-shadowed-variable
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fingerprint = exports.toDer = exports.dumpCertificate = exports.x509Date = exports.getPublicKeyFromCertificate = exports.getPublicKeyFromPrivateKey = exports.generateStaticConfig = void 0;
const assert_1 = __importDefault(require("assert"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const common_1 = require("../common");
const common2_1 = require("../common2");
const config_1 = require("../config");
const execute_openssl_1 = require("./execute_openssl");
const _env_1 = require("./_env");
function openssl_require2DigitYearInDate() {
    // istanbul ignore next
    if (!config_1.g_config.opensslVersion) {
        throw new Error("openssl_require2DigitYearInDate : openssl version is not known:" + "  please call ensure_openssl_installed(callback)");
    }
    return config_1.g_config.opensslVersion.match(/OpenSSL 0\.9/);
}
config_1.g_config.opensslVersion = "";
function generateStaticConfig(configPath, options) {
    const prePath = (options && options.cwd) || "";
    const staticConfigPath = configPath + ".tmp";
    let staticConfig = fs_1.default.readFileSync(path_1.default.join(prePath, configPath), { encoding: "utf8" });
    for (const envVar of (0, _env_1.getEnvironmentVarNames)()) {
        staticConfig = staticConfig.replace(new RegExp(envVar.pattern, "gi"), (0, _env_1.getEnv)(envVar.key));
    }
    fs_1.default.writeFileSync(path_1.default.join(prePath, staticConfigPath), staticConfig);
    return staticConfigPath;
}
exports.generateStaticConfig = generateStaticConfig;
const q = common_1.quote;
const n = common2_1.make_path;
/**
 *   calculate the public key from private key
 *   openssl rsa -pubout -in private_key.pem
 *
 * @method getPublicKeyFromPrivateKey
 * @param privateKeyFilename
 * @param publicKeyFilename
 * @param callback
 */
function getPublicKeyFromPrivateKey(privateKeyFilename, publicKeyFilename, callback) {
    (0, assert_1.default)(fs_1.default.existsSync(privateKeyFilename));
    (0, execute_openssl_1.execute_openssl)("rsa -pubout -in " + q(n(privateKeyFilename)) + " -out " + q(n(publicKeyFilename)), {}, callback);
}
exports.getPublicKeyFromPrivateKey = getPublicKeyFromPrivateKey;
/**
 * extract public key from a certificate
 *   openssl x509 -pubkey -in certificate.pem -nottext
 *
 * @method getPublicKeyFromCertificate
 * @param certificateFilename
 * @param publicKeyFilename
 * @param callback
 */
function getPublicKeyFromCertificate(certificateFilename, publicKeyFilename, callback) {
    (0, assert_1.default)(fs_1.default.existsSync(certificateFilename));
    (0, execute_openssl_1.execute_openssl)("x509 -pubkey -in " + q(n(certificateFilename)) + " > " + q(n(publicKeyFilename)), {}, callback);
}
exports.getPublicKeyFromCertificate = getPublicKeyFromCertificate;
function x509Date(date) {
    date = date || new Date();
    const Y = date.getUTCFullYear();
    const M = date.getUTCMonth() + 1;
    const D = date.getUTCDate();
    const h = date.getUTCHours();
    const m = date.getUTCMinutes();
    const s = date.getUTCSeconds();
    function w(s, l) {
        return ("" + s).padStart(l, "0");
    }
    if (openssl_require2DigitYearInDate()) {
        // for example: on MacOS , where openssl 0.98 is installed by default
        return w(Y, 2) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2) + "Z";
    }
    else {
        // for instance when openssl version is greater than 1.0.0
        return w(Y, 4) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2) + "Z";
    }
}
exports.x509Date = x509Date;
/**
 * @param certificate - the certificate file in PEM format, file must exist
 * @param callback
 */
function dumpCertificate(certificate, callback) {
    (0, assert_1.default)(fs_1.default.existsSync(certificate));
    (0, assert_1.default)(typeof callback === "function");
    (0, execute_openssl_1.execute_openssl)("x509 " + " -in " + q(n(certificate)) + " -text " + " -noout", {}, callback);
}
exports.dumpCertificate = dumpCertificate;
function toDer(certificatePem, callback) {
    (0, assert_1.default)(fs_1.default.existsSync(certificatePem));
    const certificateDer = certificatePem.replace(".pem", ".der");
    (0, execute_openssl_1.execute_openssl)("x509  " + " -outform der " + " -in " + certificatePem + " -out " + certificateDer, {}, callback);
}
exports.toDer = toDer;
function fingerprint(certificatePem, callback) {
    // openssl x509 -in my_certificate.pem -hash -dates -noout -fingerprint
    (0, assert_1.default)(fs_1.default.existsSync(certificatePem));
    (0, execute_openssl_1.execute_openssl)("x509  " + " -fingerprint " + " -noout " + " -in " + certificatePem, {}, callback);
}
exports.fingerprint = fingerprint;
